File: Test/20_local_llm/test_local_llm.py
"""
CASSIA Test 20: Local LLM Support
=================================
Tests that CASSIA correctly handles local LLM endpoints (Ollama, LM Studio, vLLM)
without requiring an API key.

Includes a full batch annotation test using real marker data.

Usage:
    python test_local_llm.py
"""

import os
import sys
import time
import socket
from pathlib import Path

# Add shared utilities to path
sys.path.insert(0, str(Path(__file__).parent.parent / "shared" / "python"))

from test_utils import (
    setup_cassia_imports,
    load_config,
    print_test_header,
    print_test_result,
    print_config_summary,
    get_test_mode
)
from result_manager import (
    create_results_dir,
    save_test_metadata,
    save_test_results,
    create_test_metadata,
    setup_logging,
    cleanup_logging
)
from fixtures import load_markers

# Setup CASSIA imports
setup_cassia_imports()


def get_ollama_model():
    """Get the first available Ollama model."""
    try:
        import requests
        response = requests.get("http://localhost:11434/api/tags", timeout=5)
        if response.status_code == 200:
            data = response.json()
            models = [m.get("name", "") for m in data.get("models", [])]
            if models:
                return models[0]
    except:
        pass
    return None


def check_ollama_running(host="localhost", port=11434, timeout=2):
    """Check if Ollama is running by attempting a socket connection."""
    try:
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.settimeout(timeout)
        result = sock.connect_ex((host, port))
        sock.close()
        return result == 0
    except:
        return False


def test_localhost_detection():
    """Test that localhost URLs are correctly detected."""
    test_cases = [
        ("http://localhost:11434/v1", True),
        ("http://localhost:1234/v1", True),
        ("http://127.0.0.1:8000/v1", True),
        ("http://LOCALHOST:11434/v1", True),  # Case insensitive
        ("https://api.deepseek.com", False),
        ("https://api.openai.com/v1", False),
    ]

    errors = []
    for url, expected in test_cases:
        is_localhost = any(x in url.lower() for x in ["localhost", "127.0.0.1"])
        if is_localhost != expected:
            errors.append(f"URL '{url}': expected {expected}, got {is_localhost}")

    return len(errors) == 0, errors


def test_api_key_bypass():
    """Test that API key is not required for localhost URLs."""
    from CASSIA import call_llm

    # Clear any existing API key
    if "CUSTOMIZED_API_KEY" in os.environ:
        del os.environ["CUSTOMIZED_API_KEY"]

    # This should NOT raise an error for localhost (even if Ollama isn't running)
    # It will fail on connection, but not on API key validation
    try:
        call_llm(
            prompt="test",
            provider="http://localhost:11434/v1",
            model="llama2",
            max_tokens=10
        )
    except ValueError as e:
        if "API key" in str(e):
            return False, ["API key was required for localhost URL"]
        # Connection error is expected if Ollama isn't running
        return True, []
    except Exception as e:
        # Connection error is fine - we just want to verify no API key error
        if "API key" in str(e).lower():
            return False, [f"API key error: {e}"]
        return True, []

    return True, []


def test_remote_requires_key():
    """Test that remote URLs still require API key."""
    from CASSIA import call_llm

    # Clear any existing API key
    if "CUSTOMIZED_API_KEY" in os.environ:
        del os.environ["CUSTOMIZED_API_KEY"]

    try:
        call_llm(
            prompt="test",
            provider="https://api.deepseek.com",
            model="deepseek-chat",
            max_tokens=10
        )
        return False, ["Remote URL should require API key but didn't raise error"]
    except ValueError as e:
        if "API key" in str(e):
            return True, []
        return False, [f"Wrong error type: {e}"]
    except Exception as e:
        if "API key" in str(e).lower():
            return True, []
        return False, [f"Wrong error: {e}"]


def test_ollama_connection():
    """Test actual Ollama connection if running."""
    from CASSIA import call_llm

    if not check_ollama_running():
        return "skipped", ["Ollama not running on localhost:11434"]

    # Try to get available models from Ollama
    try:
        import requests
        response = requests.get("http://localhost:11434/api/tags", timeout=5)
        if response.status_code == 200:
            data = response.json()
            # Use full model name including tag (e.g., "gpt-oss:20b")
            models = [m.get("name", "") for m in data.get("models", [])]
            if not models:
                return "skipped", ["Ollama running but no models installed. Run: ollama pull llama2"]
            model_to_use = models[0]  # Use first available model
            print(f"  Using Ollama model: {model_to_use}")
        else:
            return "skipped", ["Could not get Ollama model list"]
    except Exception as e:
        return "skipped", [f"Could not connect to Ollama API: {e}"]

    try:
        response = call_llm(
            prompt="Reply with exactly one word: hello",
            provider="http://localhost:11434/v1",
            model=model_to_use,
            temperature=0.1,
            max_tokens=20
        )
        if response and len(response.strip()) > 0:
            print(f"  Response: {response[:50]}...")
            return True, []
        # Empty response but connection worked - still a pass for API key bypass
        print("  Note: Empty response (model may need more time or different prompt)")
        return True, []  # Connection worked, API key bypass verified
    except Exception as e:
        # Model not found or other errors - skip gracefully
        if "not found" in str(e).lower() or "404" in str(e):
            return "skipped", [f"Model '{model_to_use}' not available: {e}"]
        return False, [f"Ollama call failed: {e}"]


def test_batch_annotation_ollama(results_dir):
    """Test full batch annotation with local Ollama - like test 02."""
    from CASSIA import runCASSIA_batch

    if not check_ollama_running():
        return "skipped", ["Ollama not running on localhost:11434"]

    model_to_use = get_ollama_model()
    if not model_to_use:
        return "skipped", ["No Ollama models available"]

    print(f"  Using Ollama model: {model_to_use}")
    print(f"  Provider: http://localhost:11434/v1")

    # Load marker data - use 2 clusters like test 02
    test_clusters = ['monocyte', 'plasma cell']
    full_df = load_markers()
    marker_df = full_df[full_df['Broad.cell.type'].isin(test_clusters)].copy()

    print(f"  Testing annotation for: {test_clusters}")
    print(f"  Marker data shape: {marker_df.shape}")

    output_name = str(results_dir / "ollama_batch_results")

    try:
        print("  Running runCASSIA_batch with local Ollama...")
        runCASSIA_batch(
            marker=marker_df,
            output_name=output_name,
            n_genes=30,
            model=model_to_use,
            temperature=0.3,
            tissue="large intestine",
            species="human",
            max_workers=1,  # Single worker for local LLM
            provider="http://localhost:11434/v1",
            validator_involvement="v1"
        )

        # Check output files
        summary_csv = Path(f"{output_name}_summary.csv")
        if summary_csv.exists():
            import pandas as pd
            results_df = pd.read_csv(summary_csv)
            clusters_annotated = len(results_df)
            print(f"  Clusters annotated: {clusters_annotated}/{len(test_clusters)}")

            if clusters_annotated == len(test_clusters):
                # Check if we got actual annotations
                if 'Final Cell Type' in results_df.columns:
                    cell_type = results_df['Final Cell Type'].iloc[0]
                    print(f"  Annotation result: {cell_type}")
                return True, []
            else:
                return False, [f"Only {clusters_annotated}/{len(test_clusters)} clusters annotated"]
        else:
            return False, ["Summary CSV not created"]

    except Exception as e:
        error_msg = str(e)
        # Handle known issues gracefully
        if "timeout" in error_msg.lower():
            return "skipped", [f"Ollama timeout (model may be too slow): {error_msg[:100]}"]
        return False, [f"Batch annotation failed: {error_msg[:200]}"]


def run_local_llm_test():
    """Run all local LLM tests."""
    print_test_header("20 - Local LLM Support Testing")

    config = load_config()
    print_config_summary(config)

    # Create results directory
    results = create_results_dir("20_local_llm", get_test_mode())
    logging_ctx = setup_logging(results['logs'])

    start_time = time.time()
    test_results = {"tests": []}
    passed = 0
    failed = 0
    skipped = 0
    errors = []

    # Test 1: Localhost detection logic
    print("\n" + "="*50)
    print("Test 1: Localhost Detection Logic")
    print("="*50)
    success, errs = test_localhost_detection()
    test_results["tests"].append({
        "name": "localhost_detection",
        "status": "passed" if success else "failed",
        "errors": errs
    })
    if success:
        print("  [OK] Localhost detection working correctly")
        passed += 1
    else:
        print(f"  [X] FAILED: {errs}")
        failed += 1
        errors.extend(errs)

    # Test 2: API key bypass for localhost
    print("\n" + "="*50)
    print("Test 2: API Key Bypass for Localhost")
    print("="*50)
    success, errs = test_api_key_bypass()
    test_results["tests"].append({
        "name": "api_key_bypass",
        "status": "passed" if success else "failed",
        "errors": errs
    })
    if success:
        print("  [OK] API key not required for localhost")
        passed += 1
    else:
        print(f"  [X] FAILED: {errs}")
        failed += 1
        errors.extend(errs)

    # Test 3: Remote still requires key
    print("\n" + "="*50)
    print("Test 3: Remote URL Requires API Key")
    print("="*50)
    success, errs = test_remote_requires_key()
    test_results["tests"].append({
        "name": "remote_requires_key",
        "status": "passed" if success else "failed",
        "errors": errs
    })
    if success:
        print("  [OK] Remote URLs correctly require API key")
        passed += 1
    else:
        print(f"  [X] FAILED: {errs}")
        failed += 1
        errors.extend(errs)

    # Test 4: Actual Ollama connection (skip if not running)
    print("\n" + "="*50)
    print("Test 4: Ollama Connection (Optional)")
    print("="*50)
    result, errs = test_ollama_connection()
    if result == "skipped":
        print(f"  [SKIP] {errs[0]}")
        skipped += 1
        test_results["tests"].append({
            "name": "ollama_connection",
            "status": "skipped",
            "errors": errs
        })
    elif result:
        print("  [OK] Successfully connected to Ollama")
        passed += 1
        test_results["tests"].append({
            "name": "ollama_connection",
            "status": "passed",
            "errors": []
        })
    else:
        print(f"  [X] FAILED: {errs}")
        failed += 1
        errors.extend(errs)
        test_results["tests"].append({
            "name": "ollama_connection",
            "status": "failed",
            "errors": errs
        })

    # Test 5: Full batch annotation with Ollama (like test 02)
    print("\n" + "="*50)
    print("Test 5: Batch Annotation with Local Ollama")
    print("="*50)
    result, errs = test_batch_annotation_ollama(results['outputs'])
    if result == "skipped":
        print(f"  [SKIP] {errs[0]}")
        skipped += 1
        test_results["tests"].append({
            "name": "batch_annotation_ollama",
            "status": "skipped",
            "errors": errs
        })
    elif result:
        print("  [OK] Batch annotation completed successfully")
        passed += 1
        test_results["tests"].append({
            "name": "batch_annotation_ollama",
            "status": "passed",
            "errors": []
        })
    else:
        print(f"  [X] FAILED: {errs}")
        failed += 1
        errors.extend(errs)
        test_results["tests"].append({
            "name": "batch_annotation_ollama",
            "status": "failed",
            "errors": errs
        })

    duration = time.time() - start_time

    # Summary
    print("\n" + "="*50)
    print("SUMMARY")
    print("="*50)
    print(f"  Passed:  {passed}")
    print(f"  Failed:  {failed}")
    print(f"  Skipped: {skipped}")
    print(f"  Duration: {duration:.2f}s")

    # Determine overall status
    if failed > 0:
        status = "failed"
    elif passed == 0:
        status = "skipped"
    else:
        status = "passed"

    # Save metadata
    metadata = create_test_metadata(
        test_name="local_llm_test",
        config=config,
        duration_seconds=duration,
        status=status,
        clusters_tested=[],
        errors=errors
    )
    metadata["test_summary"] = {
        "passed": passed,
        "failed": failed,
        "skipped": skipped
    }
    save_test_metadata(results['outputs'], metadata)
    save_test_results(results['outputs'], test_results)

    success = status == "passed"
    print_test_result(success, f"Duration: {duration:.2f}s")

    cleanup_logging(logging_ctx)
    return success


if __name__ == "__main__":
    success = run_local_llm_test()
    sys.exit(0 if success else 1)

File: Test/20_local_llm/test_local_llm.R
# CASSIA Test 20: Local LLM Support (R)
# =====================================
# Tests that CASSIA correctly handles local LLM endpoints
# without requiring an API key.
#
# Includes a full batch annotation test using real marker data.
#
# Usage:
#     Rscript test_local_llm.R

# Get script directory
get_script_dir <- function() {
  args <- commandArgs(trailingOnly = FALSE)
  file_arg <- grep("--file=", args, value = TRUE)
  if (length(file_arg) > 0) {
    return(dirname(normalizePath(sub("--file=", "", file_arg), winslash = "/")))
  }
  return(normalizePath(getwd(), winslash = "/"))
}
script_dir <- get_script_dir()

# Source shared utilities
source(file.path(script_dir, "..", "shared", "r", "test_utils.R"))
source(file.path(script_dir, "..", "shared", "r", "fixtures.R"))
source(file.path(script_dir, "..", "shared", "r", "result_manager.R"))
source(file.path(script_dir, "..", "shared", "r", "logging_manager.R"))


check_ollama_running <- function(host = "localhost", port = 11434, timeout = 2) {
  tryCatch({
    con <- socketConnection(host = host, port = port,
                           open = "r", blocking = TRUE, timeout = timeout)
    close(con)
    return(TRUE)
  }, error = function(e) {
    return(FALSE)
  })
}


get_ollama_model <- function() {
  # Get the first available Ollama model
  tryCatch({
    response <- httr::GET("http://localhost:11434/api/tags", httr::timeout(5))
    if (httr::status_code(response) == 200) {
      data <- httr::content(response, as = "parsed")
      models <- sapply(data$models, function(m) m$name)
      if (length(models) > 0) {
        return(models[1])
      }
    }
    return(NULL)
  }, error = function(e) {
    return(NULL)
  })
}


test_localhost_detection <- function() {
  # Test localhost detection in R
  test_cases <- list(
    list(url = "http://localhost:11434/v1", expected = TRUE),
    list(url = "http://localhost:1234/v1", expected = TRUE),
    list(url = "http://127.0.0.1:8000/v1", expected = TRUE),
    list(url = "https://api.deepseek.com", expected = FALSE)
  )

  errors <- list()
  for (tc in test_cases) {
    is_localhost <- grepl("localhost|127\\.0\\.0\\.1", tc$url, ignore.case = TRUE)
    if (is_localhost != tc$expected) {
      errors <- c(errors, paste0("URL '", tc$url, "': expected ", tc$expected, ", got ", is_localhost))
    }
  }

  list(success = length(errors) == 0, errors = errors)
}


test_api_key_bypass <- function() {
  # Test that setLLMApiKey works without key for localhost
  tryCatch({
    # Clear any existing key
    Sys.unsetenv("CUSTOMIZED_API_KEY")

    # This should NOT error for localhost
    setLLMApiKey(provider = "http://localhost:11434/v1")

    # Check that placeholder was set
    key <- Sys.getenv("CUSTOMIZED_API_KEY")
    if (key == "ollama") {
      return(list(success = TRUE, errors = list()))
    } else {
      return(list(success = FALSE, errors = list(paste("Expected 'ollama' placeholder, got:", key))))
    }
  }, error = function(e) {
    list(success = FALSE, errors = list(e$message))
  })
}


test_remote_requires_key <- function() {
  # Test that remote URLs still require API key
  tryCatch({
    # Clear any existing key
    Sys.unsetenv("CUSTOMIZED_API_KEY")

    # This should error
    setLLMApiKey(provider = "https://api.deepseek.com")

    # If we get here, it didn't error - that's wrong
    list(success = FALSE, errors = list("Remote URL should require API key"))
  }, error = function(e) {
    if (grepl("API key", e$message, ignore.case = TRUE)) {
      list(success = TRUE, errors = list())
    } else {
      list(success = FALSE, errors = list(paste("Wrong error:", e$message)))
    }
  })
}


test_ollama_connection <- function() {
  # Test actual Ollama connection if running
  # Note: This test verifies Ollama is reachable via HTTP API
  # The actual LLM call is tested in Test 5 (batch annotation)
  if (!check_ollama_running()) {
    return(list(status = "skipped", errors = list("Ollama not running on localhost:11434")))
  }

  model_to_use <- get_ollama_model()
  if (is.null(model_to_use)) {
    return(list(status = "skipped", errors = list("No Ollama models available")))
  }

  message("  Using Ollama model: ", model_to_use)
  message("  Ollama API is reachable and has models available")
  message("  (Full LLM call tested in Test 5 - Batch Annotation)")

  # Just verify we can set up the API key for localhost
  tryCatch({
    setLLMApiKey(provider = "http://localhost:11434/v1")
    return(list(status = TRUE, errors = list()))
  }, error = function(e) {
    return(list(status = FALSE, errors = list(paste("Failed to set up localhost:", e$message))))
  })
}


test_batch_annotation_ollama <- function(output_dir) {
  # Test full batch annotation with local Ollama - like test 02
  if (!check_ollama_running()) {
    return(list(status = "skipped", errors = list("Ollama not running on localhost:11434")))
  }

  model_to_use <- get_ollama_model()
  if (is.null(model_to_use)) {
    return(list(status = "skipped", errors = list("No Ollama models available")))
  }

  message("  Using Ollama model: ", model_to_use)
  message("  Provider: http://localhost:11434/v1")

  # Load marker data - use 2 clusters like test 02
  test_clusters <- c("monocyte", "plasma cell")
  full_df <- get_full_marker_dataframe()
  marker_df <- full_df[full_df$Broad.cell.type %in% test_clusters, ]

  message("  Testing annotation for: ", paste(test_clusters, collapse = ", "))
  message("  Marker data shape: ", nrow(marker_df), " rows x ", ncol(marker_df), " cols")

  output_name <- file.path(output_dir, "ollama_batch_results")

  tryCatch({
    message("  Running runCASSIA_batch with local Ollama...")

    # Set up API key for localhost
    setLLMApiKey(provider = "http://localhost:11434/v1")

    CASSIA::runCASSIA_batch(
      marker = marker_df,
      output_name = output_name,
      n_genes = 30,
      model = model_to_use,
      temperature = 0.3,
      tissue = "large intestine",
      species = "human",
      max_workers = 1,  # Single worker for local LLM
      provider = "http://localhost:11434/v1",
      validator_involvement = "v1"
    )

    # Check output files
    summary_csv <- paste0(output_name, "_summary.csv")
    if (file.exists(summary_csv)) {
      results_df <- read.csv(summary_csv)
      clusters_annotated <- nrow(results_df)
      message("  Clusters annotated: ", clusters_annotated, "/", length(test_clusters))

      if (clusters_annotated == length(test_clusters)) {
        # Check if we got actual annotations
        if ("Predicted.General.Cell.Type" %in% names(results_df)) {
          cell_type <- results_df$Predicted.General.Cell.Type[1]
          message("  First annotation result: ", cell_type)
        }
        return(list(status = TRUE, errors = list()))
      } else {
        return(list(status = FALSE, errors = list(paste("Only", clusters_annotated, "/", length(test_clusters), "clusters annotated"))))
      }
    } else {
      return(list(status = FALSE, errors = list("Summary CSV not created")))
    }
  }, error = function(e) {
    error_msg <- e$message
    # Handle known issues gracefully
    if (grepl("timeout", error_msg, ignore.case = TRUE)) {
      return(list(status = "skipped", errors = list(paste("Ollama timeout:", substr(error_msg, 1, 100)))))
    }
    return(list(status = FALSE, errors = list(paste("Batch annotation failed:", substr(error_msg, 1, 200)))))
  })
}


run_local_llm_test <- function() {
  print_test_header("20 - Local LLM Support Testing (R)")

  config <- load_config()
  print_config_summary(config)

  # Setup CASSIA
  tryCatch({
    setup_cassia()
    message("CASSIA R package loaded successfully")
  }, error = function(e) {
    message("Error loading CASSIA: ", e$message)
    return(FALSE)
  })

  # Create results directory
  results <- create_results_dir("20_local_llm", get_test_mode())
  start_logging(results$logs)

  start_time <- Sys.time()
  test_results <- list(tests = list())
  passed <- 0
  failed <- 0
  skipped <- 0
  errors <- list()

  # Test 1: Localhost detection
  message("\n", paste(rep("=", 50), collapse = ""))
  message("Test 1: Localhost Detection Logic")
  message(paste(rep("=", 50), collapse = ""))
  result <- test_localhost_detection()
  test_results$tests <- c(test_results$tests, list(list(
    name = "localhost_detection",
    status = if (result$success) "passed" else "failed",
    errors = result$errors
  )))
  if (result$success) {
    message("  [OK] Localhost detection working correctly")
    passed <- passed + 1
  } else {
    message("  [X] FAILED: ", paste(result$errors, collapse = ", "))
    failed <- failed + 1
    errors <- c(errors, result$errors)
  }

  # Test 2: API key bypass
  message("\n", paste(rep("=", 50), collapse = ""))
  message("Test 2: API Key Bypass for Localhost")
  message(paste(rep("=", 50), collapse = ""))
  result <- test_api_key_bypass()
  test_results$tests <- c(test_results$tests, list(list(
    name = "api_key_bypass",
    status = if (result$success) "passed" else "failed",
    errors = result$errors
  )))
  if (result$success) {
    message("  [OK] API key not required for localhost")
    passed <- passed + 1
  } else {
    message("  [X] FAILED: ", paste(result$errors, collapse = ", "))
    failed <- failed + 1
    errors <- c(errors, result$errors)
  }

  # Test 3: Remote requires key
  message("\n", paste(rep("=", 50), collapse = ""))
  message("Test 3: Remote URL Requires API Key")
  message(paste(rep("=", 50), collapse = ""))
  result <- test_remote_requires_key()
  test_results$tests <- c(test_results$tests, list(list(
    name = "remote_requires_key",
    status = if (result$success) "passed" else "failed",
    errors = result$errors
  )))
  if (result$success) {
    message("  [OK] Remote URLs correctly require API key")
    passed <- passed + 1
  } else {
    message("  [X] FAILED: ", paste(result$errors, collapse = ", "))
    failed <- failed + 1
    errors <- c(errors, result$errors)
  }

  # Test 4: Ollama connection (optional)
  message("\n", paste(rep("=", 50), collapse = ""))
  message("Test 4: Ollama Connection (Optional)")
  message(paste(rep("=", 50), collapse = ""))
  result <- test_ollama_connection()
  if (identical(result$status, "skipped")) {
    message("  [SKIP] ", result$errors[[1]])
    skipped <- skipped + 1
    test_results$tests <- c(test_results$tests, list(list(
      name = "ollama_connection",
      status = "skipped",
      errors = result$errors
    )))
  } else if (isTRUE(result$status)) {
    message("  [OK] Successfully connected to Ollama")
    passed <- passed + 1
    test_results$tests <- c(test_results$tests, list(list(
      name = "ollama_connection",
      status = "passed",
      errors = list()
    )))
  } else {
    message("  [X] FAILED: ", paste(result$errors, collapse = ", "))
    failed <- failed + 1
    errors <- c(errors, result$errors)
    test_results$tests <- c(test_results$tests, list(list(
      name = "ollama_connection",
      status = "failed",
      errors = result$errors
    )))
  }

  # Test 5: Full batch annotation with Ollama (like test 02)
  message("\n", paste(rep("=", 50), collapse = ""))
  message("Test 5: Batch Annotation with Local Ollama")
  message(paste(rep("=", 50), collapse = ""))
  result <- test_batch_annotation_ollama(results$outputs)
  if (identical(result$status, "skipped")) {
    message("  [SKIP] ", result$errors[[1]])
    skipped <- skipped + 1
    test_results$tests <- c(test_results$tests, list(list(
      name = "batch_annotation_ollama",
      status = "skipped",
      errors = result$errors
    )))
  } else if (isTRUE(result$status)) {
    message("  [OK] Batch annotation completed successfully")
    passed <- passed + 1
    test_results$tests <- c(test_results$tests, list(list(
      name = "batch_annotation_ollama",
      status = "passed",
      errors = list()
    )))
  } else {
    message("  [X] FAILED: ", paste(result$errors, collapse = ", "))
    failed <- failed + 1
    errors <- c(errors, result$errors)
    test_results$tests <- c(test_results$tests, list(list(
      name = "batch_annotation_ollama",
      status = "failed",
      errors = result$errors
    )))
  }

  duration <- as.numeric(difftime(Sys.time(), start_time, units = "secs"))

  # Summary
  message("\n", paste(rep("=", 50), collapse = ""))
  message("SUMMARY")
  message(paste(rep("=", 50), collapse = ""))
  message("  Passed:  ", passed)
  message("  Failed:  ", failed)
  message("  Skipped: ", skipped)
  message("  Duration: ", round(duration, 2), "s")

  # Determine status
  if (failed > 0) {
    status <- "failed"
  } else if (passed == 0) {
    status <- "skipped"
  } else {
    status <- "passed"
  }

  # Save metadata
  metadata <- create_test_metadata(
    test_name = "local_llm_test",
    config = config,
    duration_seconds = duration,
    status = status,
    clusters_tested = list(),
    errors = errors
  )
  metadata$test_summary <- list(passed = passed, failed = failed, skipped = skipped)
  save_test_metadata(results$outputs, metadata)

  success <- status == "passed"
  print_test_result(success, paste("Duration:", round(duration, 2), "s"))

  stop_logging()
  return(success)
}

# Run test
success <- run_local_llm_test()
quit(status = if (success) 0 else 1)
